Add read side document (#3)
EugeniyBurmistrov/resolve: docs/basics/read-side.md

# Read Models

> The reSolve framework's read side listens to events that the write side produces. Based on the events, the read side updates **Read Models**. Read Models then provide data to answer queries.

Reviewer: **Read Models**. Read Models?

Author: "Based on the events, the read side updates **Read Models**. The Read Models then provide data to answer queries."

Reviewer: Based on **these** events, the read side updates the **Read Models**, and these models provide data to the queries.

Author: done

A Read Model is defined by a set of projection functions and query resolver functions.

> - **[Projection functions](#updating-a-read-model-via-projection-functions)** build Read Model state based on incoming events.

Reviewer: "build Read Model state" a Read Model's state?

Author: fixed

> - **[Query resolvers](#resolvers)** answer queries using data from the accumulated state.

Reviewer: ...use data from the accumulated state to answer queries.

Author: done

> ReSolve also provides a special kind of Read Models that can be calculated on the fly, sent to the client and kept there up-to-date. Such Read Models are called **View Models**. Refer to the [View Model Specifics](view-model-specifics) section for more information.

Reviewer: "ReSolve also provides a special kind of Read Models that can be calculated on the fly, sent to the client and kept there up-to-date. " Explain.

Reviewer: "ReSolve also provides a special kind of Read Models that can be calculated on the fly, sent to the client and kept there up-to-date. Such Read Models are called **View Models**." You can combine these sentences. No need to mention 'a special kind of Read Models' when the next sentence tells you what they are called. Rewrite.

Author: "ReSolve also supports View Models. A View Model is a Read Model that can be built on the fly, sent to the client and kept there up-to-date."

Author: > Explain

I've tried to explain this in an earlier comment:

> on the fly?

The term "on-the-fly calculation" is widely used. The View-Model keeps being built as the events keep coming from the read side

> up-to-date

I mean, the View Model data on the client is automatically updated so it is always in the current state.

Reviewer: ReSolve also supports View Models that are automatically updated to show their current state?

# Configuring Read Models and View Models

### Configuring Read Models

All of the application's Read Models should be registered in the **config.app.js** file's **readModels** section:

```js
const appConfig = {
  ...
  readModels: [
    {
      name: 'default',
      projection: 'common/read-models/default.projection.js',
      resolvers: 'common/read-models/default.resolvers.js',
      adapter: {
        module: 'common/read-models/readmodel_adapter.module.js',
        options: {
          pathToFile: 'readmodel.db'
        }
      }
    }
  ],
  ...
}

```

In the configuration object, specify the Read Model's name and the paths to the files containing projections and resolvers. Here, you can also specify the Read Model storage adapter settings.

### Configuring View Models

> In the same way, you should register your View Models using the **viewModels** section:

Reviewer: Do not use 'using' or any other '-ing'

Author: "In the same way, you should register your View Models in the **viewModels** section:"

Reviewer: You should register your View Models in the **viewModels** section in the same way:

Author: done

```js
const appConfig = {
  ...
  viewModels: [
    {
      name: 'storyDetails',
      projection: 'common/view-models/story_details.projection.js',
      serializeState: 'common/view-models/story_details.serialize_state.js',
      deserializeState: 'common/view-models/story_details.deserialize_state.js',
      snapshotAdapter: {
        module: 'common/view-models/snapshot_adapter.module.js',
        options: {
          pathToFile: 'snapshot.db',
        }
      }
    }
  ],
  ...
}
```

In the configuration object, specify the View Model's name and the path to the file containing projection definition. You can also specify the View Model snapshot storage adapter. Use the **serializeState** and **deserializeState** options to specify paths to a View Model's serializer and deserializer functions.

# Initialize a Read Model

Each Read Model projection object should define an **Init** function that initializes the Read Model storage.

> You can add tables to the storage using the **defineTable** method:

Reviewer: Again...

Author: "You can use the **defineTable** method to add tables to the storage:"

```js
  Init: async store => {
    ...
    await store.defineTable('Comments', {
      indexes: { id: 'string' },
      fields: [
        'text',
        'parentId',
        'comments',
        'storyId',
        'createdAt',
        'createdBy',
        'createdByName'
      ]
    })
    ...
  },
```

> ReSolve provides a unified API to manage data in a storage, so this code will work with any supported storage type. The internal logic used to communicate with various DBMSs is provided by **Read Model Adapters**.

Reviewer: Why 'will work'? Why do you need to use future tense?

Reviewer: and "The internal logic used to communicate with various DBMSs is provided by **Read Model Adapters**" Why do you need to use passive voice here? Rewrite these sentences.

Author: "ReSolve provides a unified API to manage data in a storage, so this code works with any supported storage type. The internal logic that a Read Model uses to communicate with various DBMSs is provided by **Read Model Adapters**."

Reviewer: ReSolve provides a unified API to manage data in storage (this code works with any supported storage type). **Read Model Adapters** provide the internal logic a Read Model uses to communicate with DBMSs.

Author: done

> Do not hesitate to store Read Model data in a denormalized form so that your Read Models are optimized for query speed.

Reviewer: Do not hesitate?

Reviewer: "Do not hesitate to store Read Model data in a denormalized form so that your Read Models are optimized for query speed" optimized for query speed? Rewrite.

Author: We recommend that you store Read Model data in a denormalized form so that your Read Models are optimized for query performance.

Reviewer: You should store Read Model data in a denormalized form to optimize their **query performance**.

Reviewer: I'm not sure what you mean by **query performance**.

Author: Query performance means how fast a client can perform a data query and receive an answer from the server. For example:
https://docs.mongodb.com/manual/tutorial/optimize-query-performance-with-indexes-and-projections/
https://solutioncenter.apexsql.com/how-to-optimize-sql-server-query-performance/

# Updating a Read Model via Projection Functions

A projection function is used to accumulate the event data in a **Read Model storage**. Each projection function takes the storage object and event information, including the aggregateID, timestamp and payload.

> You can communicate with the store using the standard API. The code sample below demonstrates a typical Read Model projection function implementation:

Reviewer: -ing

Author: "...and event information. The event information includes the aggregateID, timestamp and payload."

```js
[STORY_COMMENTED]: async (
  store, { aggregateId, timestamp, payload: { parentId, userId, userName, commentId, text } }
) => {
  const comment = { id: commentId, text, parentId, comments: [], storyId: aggregateId,
    createdAt: timestamp, createdBy: userId, createdByName: userName }

  await store.insert('Comments', comment)
  await store.update(
    'Stories',
    { id: aggregateId },
    { $inc: { commentCount: 1 } }
  )
}
...
```

> The data from a populated store is then used by a [resolver](#resolver) to prepare final data samples in response to data requests.

Reviewer: Again, unnecessary passive voice. Rewrite.

Author: "A [resolver](#resolvers) then uses the data from the store to prepare final data samples in response to data requests."

Reviewer: A [resolver](#resolvers) then uses the data from the store to prepare final data samples for data requests.

Author: done

> You can force the framework to re-populate the store using events starting from the beginning of the history by deleting the Read Model storage. This can be useful in the development environment and when deploying an updated version of the application.

Reviewer: '-ing' - Rewrite.

Author: "If you delete the Read Model storage, this will force the framework to re-populate the store based on all events from the beginning of the history. This can be useful in the development environment and when you deploy an updated version of the application."

Reviewer: If you delete the Read Model storage, the framework re-populates the store based on all the events.

Reviewer: What do you mean "This can be useful in the development environment and when you deploy an updated version of the application"? Why 'useful'?

Author: When you develop an application and run it on your machine - that's the development environment. By "deploy an updated version" I mean that you modify your application and deploy this new version to your web server (the production environment). In these two cases, you may want to rebuild your Read Models from scratch, which you can achieve by deleting the storage. This is why I say that deleting the storage can be useful.

Author: > If you delete the Read Model storage, the framework re-populates the store based on all the events.

Done

> Note that reSolve does not limit you on what logic you can use in a projection function implementation as long as it helps you prepare data required to answer queries. Depending on your requirements, you can perform SQL queries, update Elastic Search indexes, write arbitrary data to files, etc.

Reviewer: "Note that reSolve does not limit you on what logic you can use in a projection function implementation as long as it helps you prepare data required to answer queries." What are you trying to say here?

Author: There is a misconception that a projection function can only take information from event objects, process it and write the result to a Read Model storage. In this sentence we are trying to say that no, you actually can write whatever you want in the function's implementation: take information from external sources, write data to logs or databases, etc. The only thing that you should care about is that your projection function does what a projection function should do - it prepares the data required to answer queries.

Reviewer: You should rewrite this.

Author: Note that you can add any additional logic to a projection function. For instance, you can perform SQL queries, update Elastic Search indexes, write arbitrary data to files, etc.

Reviewer: Note that you can add additional logic to a projection function. For instance, you can perform SQL queries, update Elastic Search indexes, write arbitrary data to files, etc.

# Resolvers

> A **Read Model resolver** is the the part of a Read Model that handles data requests. A resolver function receives the store and request parameters. Based on the parameters, the resolver function pulls the required data from the store and processes it to prepare the response object.

Reviewer: "A **Read Model resolver** is the the part of a Read Model that handles data requests." the the? and why do you repeat 'Read Model'? Rewrite.

Author: "A **resolver** is the part of a Read Model that handles data requests."

The code sample below demonstrate a typical Read Model implementation:

```js
comments: async (store, { first, offset }) => {
  const skip = first || 0
  const comments = await store.find(
    'Comments',
    {},
    null,
    { createdAt: -1 },
    skip,
    skip + offset
  )
  return Array.isArray(comments) ? comments : []
}
```

> To learn how to send a request to a Read Model resolver, refer to the [Query a Read Model](#query_a_read_model) section.

Reviewer: 'To learn' -> for more information or for more details. refer to the [Query a Read Model](#query_a_read_model) section for more information on how to send a request to a Read Model resolver.

Author: done

Author: "To learn" and "for more information" are not equivalent in this context. We cannot say "for _more_ information" because we did not provide any information at all on this topic in the current section.

What I am trying to say is: "in this section you learned how what is a Read Model and how to create is; to also learn how to query it, see the Query a Read Model"

Reviewer: You do not 'learn' something from a section. Documents provide information about how things work or what they do, or provide instructions on how to accomplish a certain task.

Author: "refer to the [Query a Read Model](#query_a_read_model) section for information on how to..."

# View Model Specifics

> **View Models** are a special kind of Read Models. They are queried based on aggregate ID and and can automatically provide updates to Redux state on the client. View Models are defined in a special isomorphic format so their code can also be used on the client side to provide reducer logic.

Reviewer: You repeatedly mention 'a special kind of'. Why? This is a special kind of Read Model, that is a special kind of Read Model? and it cannot be 'a' Models.

Author: People keep perceiving them as a separate entity with some magical qualities. We added the 'special kind of' part because people keep asking why we have two types of 'models' and why they should choose one over the other. At the same time, saying that they are just Read Models is not exactly accurate, because they are created and configured differently, at least at the current state of the development.

Reviewer: Then write that. 'a special kind of Read Models' does not mean anything without explaining what it is. You should explain why they are different instead of just saying they are 'special'.

Author: The whole point of this section is to explain what's special about them.  We added the first sentence to point out the logical relation between these two concepts.

Use View Models in the following scenarios:

- To create aggregate-centric views. Such views request relatively portions of data based on aggregate IDs.
- To create reactive components, whose state is kept up-to date on the client.

> A View Model's projection function recieves a state and an event object, and returns an updated state. A projection function runs for every event with the specified aggregate ID from the beginning of the history on every request so it is important to keep View Models small. You can also store snapshots of the View Model state to optimize system resource consumption.

Reviewer: recieves?

Author: receives
fixed

The code sample below demonstrate a typical View Model projection function:

```js
[SHOPPING_ITEM_CREATED]: (state, { payload: { id, text } }) => ({
  ...state,
  list: [
    ...state.list,
    {
      id,
      text,
      checked: false
    }
  ]
}),
```

> To learn how to send a request to a View Model projection, refer to the [Query a View Model](#query_a_view_model) section.

Reviewer: Same here

Author: See my answer above

> Note that a View model does not use the Read Model store in any way.

Reviewer: Note that a View model does not use the Read Model store in any way? in any way? Explain.

Reviewer: Why do you need to add 'in any way'?

Author: Removed 'in any way'

| **resolver**  | The name of a [resolver defined in the Read Model](#resolvers)                         |

The request body should have the `application/json` content type and the following structure:

```js
{
  param1: value1,
  param2: value2,
  // ...
  paramN: valueN
}
```

The object contains the parameters that the resolver accepts.

##### Example

Use the following command to get 3 users from the [with-saga](../examples/with-saga) example.

```sh
curl -X POST \
-H "Content-Type: application/json" \
-d "{\"page\":0, \"limit\":3}" \
"http://localhost:3000/api/query/default/users"
```

### Query a View Model

You can query a View Model from the client side by sending a POST request to the following URL:

```
http://{host}:{port}/api/query/{viewModel}/{aggregateIds}
```

##### URL Parameters

| Name         | Description                                                                                                 |
| ------------ | ----------------------------------------------------------------------------------------------------------- |
| viewModel    | The View Model name as defined in [config.app.js](../examples/shopping-list/config.app.js)                  |

> | aggregateIds | The comma-separated list of Aggregate IDs to include into the View Model. Use `*` to include all Aggregates |

Reviewer: to include into? Why 'into'?

Author: changed to 'in'

# Read Models

> The Read Side of the reSolve framework listens to events produced by the write side and updates **Read Models**. Read Models are then used to answer queries. 

Reviewer: "produced by" Why are you using passive voice here? We've discussed this before.

Reviewer: "listens to events" listens?

Reviewer: "The Read Side of the reSolve framework" Why are you not using a possesive here? We've discussed this before.

Reviewer: updates **Read Models** -> updates **the** Read Models

> A Read Model is defined using functions of the following two kinds:

Reviewer: two kinds?

Reviewer: The following functions are used to define a Read Model:

> * **[Projection functions](#updating-a-read-model-via-projection-functions)** - Applies events to accumulate state.

Reviewer: to accumulate state?

* **[Query resolver](#resolvers)** - Answers queries based on the accumulated state.

> ReSolve also provides a special kind of Read Models that can be calculated on the fly, sent to the client and kept there up-to-date. Such Read Models are called **View Models**. Refer to the [View Model Specifics](view-model-specifics) section for more information.

Reviewer: a special kind?

Reviewer: on the fly?

Reviewer: kept there up-to-date?

Author: > a special kind?

I don't see what's wrong with this, could you elaborate?

Author: > on the fly?

The term "on-the-fly calculation" is widely used. The View-Model is calculated as the events keep coming from the read side

Author: > up-to-date

I mean, the View Model data on the client is automatically updated so it is always in the current state.

Reviewer: ```suggestion
ReSolve also provides Read Models called **View Models** that can be calculated on the fly, sent to the client, and are automatically updated. Refer to the [View Model Specifics](view-model-specifics) section for more information.
```

# Configuring Read Models and View Models
### Configuring Read Models

> All application's Read Models should be registered in the **config.app.js** file's **readModels** section:

Reviewer: All application's? An application's?

Reviewer: Rewrite this doc.

``` js
const appConfig = {
  ...
  readModels: [
    {
      name: 'default',
      projection: 'common/read-models/default.projection.js',
      resolvers: 'common/read-models/default.resolvers.js',
      adapter: {
        module: 'common/read-models/readmodel_adapter.module.js',
        options: {
          pathToFile: 'readmodel.db'
        }
      }
    }
  ],
  ...
}

```

> In the configuration object, specify the Read Model's name and the paths to the projection's and resolvers' definitions. Here, you can also specify the storage adapter settings for the Read Model.

Reviewer: projection's?

Reviewer: Again, "for the"
